src/mds/Server.cc

```diff
@@ -3321,6 +3321,14 @@ CInode* Server::prepare_new_inode(MDRequestRef& mdr, CDir *dir, inodeno_t useino
     // xattrs on new inode?
     auto _xattrs = CInode::allocate_xattr_map();
     decode_noshare(*_xattrs, p);
+
+    // The "alternate_name" is not a real xattr,
+    // will remove it from the xattrs map.
+    auto it = _xattrs->find("alternate_name");
+    if (it != _xattrs->end()) {
+      dn->set_alternate_name(std::move(it->second));
+      _xattrs->erase(it);
+    }
```

Comment:
@jtlayton couldn't we transmit the alternate_name by adding it to the message payload? This is a terrible hack.

Reply:
> @jtlayton couldn't we transmit the alternate_name by adding it to the message payload? This is a terrible hack.

Yes. This does seem like a nasty hack.